File: templates/Uwp/_comp/_shared/Page.Ink.AddUndoRedoService._VB/Services/Ink/InkUndoRedoService.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports Param_RootNamespace.EventHandlers.Ink
Imports Param_RootNamespace.Services.Ink.UndoRedo
Imports Windows.UI.Input.Inking

Namespace Services.Ink
    Public Class InkUndoRedoService
        Private ReadOnly _strokeService As InkStrokesService
        Private undoStack As Stack(Of IUndoRedoOperation) = New Stack(Of IUndoRedoOperation)()
        Private redoStack As Stack(Of IUndoRedoOperation) = New Stack(Of IUndoRedoOperation)()

        Public Sub New(inkCanvas As InkCanvas, strokeService As InkStrokesService)
            _strokeService = strokeService
            AddHandler _strokeService.StrokesCollected, AddressOf StrokeService_StrokesCollected
            AddHandler _strokeService.StrokesErased, AddressOf StrokeService_StrokesErased
            AddHandler _strokeService.MoveStrokesEvent, AddressOf StrokeService_MoveStrokesEvent
            AddHandler _strokeService.CutStrokesEvent, AddressOf StrokeService_CutStrokesEvent
            AddHandler _strokeService.PasteStrokesEvent, AddressOf StrokeService_PasteStrokesEvent
        End Sub

        Public Event UndoEvent As EventHandler(Of EventArgs)

        Public Event RedoEvent As EventHandler(Of EventArgs)

        Public Event AddUndoOperationEvent As EventHandler(Of EventArgs)

        Public Sub Reset()
            undoStack.Clear()
            redoStack.Clear()
            RaiseEvent AddUndoOperationEvent(Me, EventArgs.Empty)
        End Sub

        Public ReadOnly Property CanUndo As Boolean
            Get
                Return undoStack.Any()
            End Get
        End Property

        Public ReadOnly Property CanRedo As Boolean
            Get
                Return redoStack.Any()
            End Get
        End Property

        Public Sub Undo()
            If Not CanUndo Then
                Return
            End If

            RemoveHandler _strokeService.MoveStrokesEvent, AddressOf StrokeService_MoveStrokesEvent
            Dim element = undoStack.Pop()
            element.ExecuteUndo()
            redoStack.Push(element)
            AddHandler _strokeService.MoveStrokesEvent, AddressOf StrokeService_MoveStrokesEvent
            RaiseEvent UndoEvent(Me, EventArgs.Empty)
        End Sub

        Public Sub Redo()
            If Not CanRedo Then
                Return
            End If

            RemoveHandler _strokeService.MoveStrokesEvent, AddressOf StrokeService_MoveStrokesEvent
            Dim element = redoStack.Pop()
            element.ExecuteRedo()
            undoStack.Push(element)
            AddHandler _strokeService.MoveStrokesEvent, AddressOf StrokeService_MoveStrokesEvent
            RaiseEvent RedoEvent(Me, EventArgs.Empty)
        End Sub

        Public Sub AddOperation(operation As IUndoRedoOperation)
            If operation Is Nothing Then
                Return
            End If

            undoStack.Push(operation)
            redoStack.Clear()
            RaiseEvent AddUndoOperationEvent(Me, EventArgs.Empty)
        End Sub

        Private Sub StrokeService_StrokesCollected(sender As Object, e As InkStrokesCollectedEventArgs)
            AddStrokesOperation(e.Strokes)
        End Sub

        Private Sub StrokeService_StrokesErased(sender As Object, e As InkStrokesErasedEventArgs)
            RemoveStrokesOperation(e.Strokes)
        End Sub

        Private Sub StrokeService_CutStrokesEvent(sender As Object, e As CopyPasteStrokesEventArgs)
           RemoveStrokesOperation(e.Strokes)
        End Sub

        Private Sub StrokeService_PasteStrokesEvent(sender As Object, e As CopyPasteStrokesEventArgs)
            AddStrokesOperation(e.Strokes)
        End Sub

        Private Sub StrokeService_MoveStrokesEvent(sender As Object, e As MoveStrokesEventArgs)
            Dim operation = New MoveStrokesUndoRedoOperation(e.Strokes, e.StartPosition, e.EndPosition, _strokeService)
            AddOperation(operation)
        End Sub

        Private Sub AddStrokesOperation(strokes As IEnumerable(Of InkStroke))
            Dim operation = New AddStrokeUndoRedoOperation(strokes, _strokeService)
            AddOperation(operation)
        End Sub

        Private Sub RemoveStrokesOperation(strokes As IEnumerable(Of InkStroke))
            Dim operation = New RemoveStrokeUndoRedoOperation(strokes, _strokeService)
            AddOperation(operation)
        End Sub
    End Class
End Namespace

File: code/TemplateStudioForUWP/Templates/Uwp/_comp/_shared/Page.Ink.AddUndoRedoService/Services/Ink/InkUndoRedoService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Windows.UI.Xaml.Controls;
using Windows.UI.Input.Inking;
using Param_RootNamespace.EventHandlers.Ink;
using Param_RootNamespace.Services.Ink.UndoRedo;

namespace Param_RootNamespace.Services.Ink
{
    public class InkUndoRedoService
    {
        private readonly InkStrokesService _strokeService;

        private Stack<IUndoRedoOperation> undoStack = new Stack<IUndoRedoOperation>();
        private Stack<IUndoRedoOperation> redoStack = new Stack<IUndoRedoOperation>();

        public InkUndoRedoService(InkCanvas inkCanvas, InkStrokesService strokeService)
        {
            _strokeService = strokeService;
            _strokeService.StrokesCollected += StrokeService_StrokesCollected;
            _strokeService.StrokesErased += StrokeService_StrokesErased;
            _strokeService.MoveStrokesEvent += StrokeService_MoveStrokesEvent;
            _strokeService.CutStrokesEvent += StrokeService_CutStrokesEvent;
            _strokeService.PasteStrokesEvent += StrokeService_PasteStrokesEvent;
        }

        public event EventHandler<EventArgs> UndoEvent;

        public event EventHandler<EventArgs> RedoEvent;

        public event EventHandler<EventArgs> AddUndoOperationEvent;

        public void Reset()
        {
            undoStack.Clear();
            redoStack.Clear();
            AddUndoOperationEvent?.Invoke(this, EventArgs.Empty);
        }

        public bool CanUndo => undoStack.Any();

        public bool CanRedo => redoStack.Any();

        public void Undo()
        {
            if (!CanUndo)
            {
                return;
            }

            _strokeService.MoveStrokesEvent -= StrokeService_MoveStrokesEvent;

            var element = undoStack.Pop();
            element.ExecuteUndo();
            redoStack.Push(element);

            _strokeService.MoveStrokesEvent += StrokeService_MoveStrokesEvent;
            UndoEvent?.Invoke(this, EventArgs.Empty);
        }

        public void Redo()
        {
            if (!CanRedo)
            {
                return;
            }

            _strokeService.MoveStrokesEvent -= StrokeService_MoveStrokesEvent;

            var element = redoStack.Pop();
            element.ExecuteRedo();
            undoStack.Push(element);

            _strokeService.MoveStrokesEvent += StrokeService_MoveStrokesEvent;
            RedoEvent?.Invoke(this, EventArgs.Empty);
        }

        public void AddOperation(IUndoRedoOperation operation)
        {
            if (operation == null)
            {
                return;
            }

            undoStack.Push(operation);
            redoStack.Clear();
            AddUndoOperationEvent?.Invoke(this, EventArgs.Empty);
        }

        private void StrokeService_StrokesCollected(object sender, InkStrokesCollectedEventArgs e) => AddStrokesOperation(e.Strokes);

        private void StrokeService_StrokesErased(object sender, InkStrokesErasedEventArgs e) => RemoveStrokesOperation(e.Strokes);

        private void StrokeService_CutStrokesEvent(object sender, CopyPasteStrokesEventArgs e) => RemoveStrokesOperation(e.Strokes);

        private void StrokeService_PasteStrokesEvent(object sender, CopyPasteStrokesEventArgs e) => AddStrokesOperation(e.Strokes);

        private void StrokeService_MoveStrokesEvent(object sender, MoveStrokesEventArgs e)
        {
            var operation = new MoveStrokesUndoRedoOperation(e.Strokes, e.StartPosition, e.EndPosition, _strokeService);
            AddOperation(operation);
        }

        private void AddStrokesOperation(IEnumerable<InkStroke> strokes)
        {
            var operation = new AddStrokeUndoRedoOperation(strokes, _strokeService);
            AddOperation(operation);
        }

        private void RemoveStrokesOperation(IEnumerable<InkStroke> strokes)
        {
            var operation = new RemoveStrokeUndoRedoOperation(strokes, _strokeService);
            AddOperation(operation);
        }
    }
}